homeassistant/config_entries.py

```diff
@@ -331,6 +342,7 @@ async def async_setup(
 
         if integration is None:
```

Comment:
`setup_again` used to pass  `integration=integration` and now does not, so `loader.async_get_integration` gets invoked each time. Does that add meaningful overhead when retrying setup? I'm wondering 
(1) if the `integration` parameter passed in here should just be removed entirely given its only used in one other place that immediately calls `async_get_integration` right before anyway.
(2) if this should just have a guard of only loading when `self._integration_for_domain` is `None`

Reply:
Re-reading `async_forward_entry_setup` I may not understand what it is doing. So maybe there are two cases here: integration is None which is the default, then integration where its another domain, hence why this only runs when no integration is passed in. What i'm unsure about is how retry works on the latter case when an integration is passed in.

Reply:
> `setup_again` used to pass `integration=integration` and now does not, so `loader.async_get_integration` gets invoked each time. Does that add meaningful overhead when retrying setup? I'm wondering 

> (1) if the `integration` parameter passed in here should just be removed entirely given its only used in one other place that immediately calls `async_get_integration` right before anyway. 

We need to be able to override `integration` when loading platforms like `sensor` etc.

Ie `nest.sensor`

> (2) if this should just have a guard of only loading when `self._integration_for_domain` is `None`

We can't do that because each platform needs the integration for the platform and not the integration for the `domain` i.e. `nest` itself.

Reply:
> Re-reading `async_forward_entry_setup` I may not understand what it is doing. So maybe there are two cases here: integration is None which is the default, then integration where its another domain, hence why this only runs when no integration is passed in. 

> What i'm unsure about is how retry works on the latter case when an integration is passed in.

There is no retry for anything but the main integration `domain`, once we pass control to `async_forward_entry_setup` we are past the point of being able to raise `ConfigEntryXXXXX` and control the setup flow.

Reply:
Hopefully that all made sense. Its a bit confusing because of `integration` being passed around and how the same code is used for the platform setup via `async_forward_entry_setup`

Reply:
If it helps: in `async_forward_entry_setup` the `domain` is actually the `platform` like `sensor`, etc.

Reply:
Yes, that flow all makes sense. I think what is hard to reason about is the assumptions/expectations based on the exception thrown, if I understand right. That is, if `nest.sensor` happens to throw `ConfigEntryNotReady` then it think it may retry with the wrong domain? I realize that would be against our design standard, however, it seems like having some invariants added could help. (e.g. bake in assumptions about what exceptions are thrown and handled vs not allowed when in forward entry steup)

Reply:
If `nest.sensor` threw `ConfigEntryNotReady` or `ConfigEntryAuthFailed` or `ConfigEntryError` the code would not behave well.  It would be nice to handle these differently and blow up with an implementation error instead `if self.domain != integration.domain:` but thats but this is an existing problem which should be solved in another PR if there is a desire to do it.

Reply:
There is so much branching here that it likely makes sense to create a `async_setup_platform` or modify `async_forward_entry_setup` to do a subset of `async_setup`....but thats a bit off into the weeds.

Reply:
I may be confused about whats happening here -- is this loading `sensor`? Then the entity platform/component ends up calling `nest.sensor`? I was looking where `PlatformNotReady` is handled, then remembered its not here.

I think the easiest way to make the mistake of the wrong exception being thrown is by using an update coordinator's methods that throw exceptions for retrying setup though I only found one example:
https://github.com/home-assistant/core/blob/f4f78cf00076af95be427cf6f6ddbcc8fefcb259/homeassistant/components/filesize/sensor.py#L75 -- though given the above not sure if that really even runs here.

Reply:
> I may be confused about whats happening here -- is this loading `sensor`? Then the entity platform/component ends up calling `nest.sensor`? I was looking where `PlatformNotReady` is handled, then remembered its not here.
> 

`PlatformNotReady` goes though the `entity_platform` setup path https://github.com/home-assistant/core/blob/f4f78cf00076af95be427cf6f6ddbcc8fefcb259/homeassistant/helpers/entity_platform.py#L372

> I think the easiest way to make the mistake of the wrong exception being thrown is by using an update coordinator's methods that throw exceptions for retrying setup though I only found one example:
> 
> https://github.com/home-assistant/core/blob/f4f78cf00076af95be427cf6f6ddbcc8fefcb259/homeassistant/components/filesize/sensor.py#L75
> 
> -- though given the above not sure if that really even runs here.

That's definitely a mistake in the existing code.

Reply:
https://github.com/home-assistant/core/issues/99518

Reply:
> There is so much branching here that it likely makes sense to create a `async_setup_platform` or modify `async_forward_entry_setup` to do a subset of `async_setup`....but thats a bit off into the weeds.

Agree that copying code could be simpler.

Reply:
Added to my backlog